blog/2025-11-12-best-practises-for-writing-promises/index.mdx

```diff
@@ -0,0 +1,559 @@
+```
+
+You should see that the following has been added to the `promise.yaml`
+
+```yaml
+  api:
+    apiVersion: apiextensions.k8s.io/v1
+    kind: CustomResourceDefinition
+    metadata:
+      creationTimestamp: null
+      name: rediss.mygroup.org
+    spec:
+      group: mygroup.org
+      names:
+        kind: Redis
+        plural: rediss
+        singular: redis
+      scope: Namespaced
+      versions:
+      - name: v1alpha1
+        schema:
+          openAPIV3Schema:
+            properties:
+              spec:
+                properties:
+                  size:
+                    type: string
+                type: object
+            type: object
+        served: true
+        storage: true
+```
+
+We can now add a workflow that uses to this property. Again, we can use the kratix cli - this time with the `add container`
+command:
+
+```bash
+kratix add container resource/configure/instance-configure --image ghcr.io/syntasso/redis-dependencies:v0.0.1 --language go
+```
+
+This will have updated the `promise.yaml` to add a `resource.configure` workflow and have bootstrapped a Docker image that can be
+built to run as part of this workflow. As we has specified that we want to use the [kratix-go sdk](https://pkg.go.dev/github.com/syntasso/kratix-go) for this Promise via the `--language`
+flag, we also have a `pipeline.go` file. Update the script to contain the following:
+
+```go
+package main
+
+import (
+	"fmt"
+
+	kratix "github.com/syntasso/kratix-go"
+)
+
+func main() {
+	sdk := kratix.New()
+	resource, _ := sdk.ReadResourceInput()
+	name := resource.GetName()
+	size, err := resource.GetValue("spec.size")
+	if err != nil {
+		log.Fatalf("failed fetch 'size' property from request: %v", err)
+	}
+	
+	var replicas int
+	if size == "small" {
+		replicas = 1
+	}
+	
+	if size == "large" {
+		replicas = 3
+	}
+	
+	failoverTemplate := `apiVersion: databases.spotahome.com/v1
+kind: RedisFailover
+metadata:
+  name: %s
+  namespace: default
+spec:
+  redis:
+    imagePullPolicy: IfNotPresent
+    replicas: %d
+    customConfig:
+      - "maxclients 100"
+      - "hz 50"
+      - "timeout 60"
+      - "tcp-keepalive 60"
+      - "client-output-buffer-limit normal 0 0 0"
+      - "client-output-buffer-limit slave 1000000000 1000000000 0"
+      - "client-output-buffer-limit pubsub 33554432 8388608 60"
+    resources:
+      limits:
+        cpu: 400m
+        memory: 500Mi
+      requests:
+        cpu: 100m
+        memory: 100Mi
+  sentinel:
+    imagePullPolicy: IfNotPresent
+    replicas: %d
+    customConfig:
+      - "down-after-milliseconds 2000"
+      - "failover-timeout 3000"
+    resources:
+      limits:
+        memory: 100Mi
+      requests:
+        cpu: 100m
+`
+		redisContent := []byte(fmt.Sprintf(failoverTemplate, name, replicas, replicas))
+	
+		sdk.WriteOutput("redis_failover.yaml", redisContent)
+}
+```
+
+We’re going to execute this script when creating and updating a resource request. It will:
```

Comment:
```suggestion
We're going to execute this script when creating and updating a resource request. It will:
```

Will the go files be executed as a script or compiled into a binary? It feels slightly odd to have it called as a script, so if we're executing the go files as scripts it'd be useful to mention it in an admonition.